scripts/add_strelka2_fields.py

```diff
@@ -0,0 +1,182 @@
+    except ValueError:
+        if len(set(sgt)) == 2:
+            tumor_gt = (0, 1)
+        elif len(set(sgt)) == 1:
+            if sgt[0] == ref:
+                tumor_gt = (0, 0)
+            else:
+                tumor_gt = (1, 1)
+        else:
+            raise IOError('Unhandled Strelka2 SGT value %s' % sgt)
+        
+    return tumor_gt
+
+def get_normal_GT(info_nt):
+    """ Produce standard GT for a Strelka2 normal sample
+
+        Args:
+            info_nt: value of info['NT'] for pysam VariantRecord object
+
+        Returns:
+            genotype (tuple): e.g. (0, 1)
+    """
+    return name_to_gt(info_nt)
+
+def get_AD(vcf_sample, ref, alts):
+    """ Produce standard AD for a Strelka2 sample
+        
+        Args:
+            vcf_sample (pysam VariantRecord sample object)
+            ref (str): value of ref for pysam VariantRecord object
+            alts (tuple): value of alts for pysam VariantRecord object
+        Returns:
+            allele depths (tuple): (ref, alt)
+    
+        From strelka2 docs:
+        SNPs:
+            refCounts = Sum values FORMAT column $REF + “U”
+                (e.g. if REF="A" then use the value in FORMAT/AU)
+            altCounts = Sum values FORMAT column $ALT + “U”
+                (e.g. if ALT="T" then use the value in FORMAT/TU)
+            Indels:
+                tier1RefCounts = Sum values from FORMAT/TAR
+                tier1AltCounts = Sume values from FORMAT/TIR
+    """
+    # only indels should have TIR field
+    if 'TIR' in vcf_sample:
+        ref_ct = sum(vcf_sample['TAR'])
+        alt_ct = sum(vcf_sample['TIR'])
+    else: # SNP case
+        snp_ref_key = ref[0] + "U"
+        snp_alt_key = alts[0][0] + "U"
+        ref_ct = sum(vcf_sample[snp_ref_key])
+        alt_ct = sum(vcf_sample[snp_alt_key])
+
+    return (ref_ct, alt_ct)
+
+def get_AF(ad):
+    """ Calculate allele frequency from allele depths 
+        
+        Args:
+             ad (tuple): (ref, alt) depths
+
+        Returns:
+             alt allele frequency (float) 
+    """
+    ref_counts, alt_counts = ad
+    total_counts = ref_counts + alt_counts
+    if not total_counts:
+        return 0.0
+    return alt_counts / total_counts
+
+def create_mod_vcf(output_path, input_path, tumor_name, normal_name):
+    """ Create a new VCF in which standard FORMAT fields have been
+            calculated and added=based on those provided natively by Strelka2
+
+        Args:
+            output_path (str): path to the VCF being created
+            input_path (str): path to the Strelka2 VCF
+            tumor_name (str): name of tumor sample in input
+            normal_name (str) name of normal sample in input
+
+        Raises:
+            IOError if a sample name in the input is neither tumor_name
+                nor normal_name
+    """
+
+    input_vcf = pysam.VariantFile(input_path, 'r')
+    output = pysam.VariantFile(output_path, 'w', header=input_vcf.header)
```

Comment:
I think this is missing a step where you add the updated header, for instance, from my old dev script:
```
strelka2_in.header.add_meta('FORMAT', items=[('ID','GT'), ('Number',1), ('Type','String'), ('Description', 'Genotype converted for cross-compatibility using bcbio method')])
strelka2_in.header.add_meta('FORMAT', items=[('ID','AF'), ('Number', '.'), ('Type', 'Float'), ('Description', 'Allele frequency, as recommended by strelka2 docs: <ALT>U/<REF>U+<ALT>U (somatic snps), TIR/TIR+TAR (somatic indels)')])
strelka2_in.header.add_meta('FORMAT', items=[('ID','AD'), ('Number', 'R'), ('Type', 'Integer'), ('Description', 'Allelic depths for the ref and alt alleles in the order listed. Added in post for compatibility')])
```
When I run this script I get error message:
```
Traceback (most recent call last):
  File "/home/ubuntu/tools/kf-somatic-workflow/scripts/add_strelka2_fields.py", line 181, in <module>
    create_mod_vcf(output_vcf_path, args.strelka2_vcf, args.tumor_name, args.normal_name)
  File "/home/ubuntu/tools/kf-somatic-workflow/scripts/add_strelka2_fields.py", line 134, in create_mod_vcf
    sample['GT'] = get_tumor_GT(record.info['SGT'], record.ref)
  File "pysam/libcbcf.pyx", line 3463, in pysam.libcbcf.VariantRecordSample.__setitem__
  File "pysam/libcbcf.pyx", line 844, in pysam.libcbcf.bcf_format_set_value
KeyError: 'unknown format: GT'
```
Otherwise, everything else seems to be in order

Reply:
The header of the Strelka2 VCF I started with already contained these FORMAT fields, and indeed it turns out that if they're already there you can't really change them with pysam. Would you be able to point me to the VCF you used as input to this script? I was using this one as indicated in the test files since it had the `final` tag: https://cavatica.sbgenomics.com/u/zhangb1/kf-somatic-tools-test/files/5ffdf91de4b0a6d32f329e75/

However I see now that it looks like the header sections in that VCF for these fields were written by you, so I'm guessing that this is not the one I want to use. I got it from here: https://cavatica.sbgenomics.com/u/zhangb1/kf-somatic-tools-test/files/#q?path=CONSENSUS_UPDATE_TEST&page=1&label_id=NORMED

Do I just want to use the one with the `normed` and `mnp` tags?

Reply:
Actually, as it stands, you'd run your script pre-mnp building, so this one is probably best to test on: https://cavatica.sbgenomics.com/u/zhangb1/kf-somatic-tools-test/files/5ffdf91ce4b0a6d32f329e6b/

Reply:
Ok thanks! It works now, just use that Strelka2 VCF instead of the one I indicated in test files above.